### PR TITLE
feat(#101): CachedModelSource.java for faster ml tests

### LINKED ISSUE #101: ModelSourceFileSystemTest.java:37-40: Enable ML tests...
The puzzle `1-e3e1b034` from #1 has to be resolved: 

https://github.com/volodya-lombrozo/jtcop/blob/28565add971c0a08076d927da5641a9b70a2e7f8/src/test/java/com/github/lombrozo/testnames/rules/ml/ModelSourceFileSystemTest.java#L37-L40

The puzzle was created by @volodya-lombrozo on 16-Apr-23. 

Estimate: 90 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by h1alexbel**
In this pull I've implemented `CachedModelSource.java` to cache model download in `ModelSourceInternet.java`. Enabled all unit tests related to ml.

closes #101

---

## PR-Codex overview
The focus of this PR is to update exception handling and introduce caching for model sources in the ML module.

### Detailed summary
- Updated exception handling in `ModelSource.java`
- Introduced caching in `CachedModelSource.java` and `CachedModelFs.java`
- Modified test classes to accommodate changes

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by h1alexbel** [comment]
@volodya-lombrozo take a look, please

**Review by volodya-lombrozo — COMMENTED** [review]
@h1alexbel The tests are rather fast indeed. Although I have some doubts about the following line in this test:

```java
RulePresentSimpleMlTest.model = new POSTaggerME(new ModelSourceInternet().model());
```
It seems we need to download this model each time when we start testing. That in turn might be rather time consuming. Please, correct me if I'm wrong (I haven't checked.)

**Comment by h1alexbel** [comment]
@volodya-lombrozo yes, definitely. I didn't check the speed of code in `@BeforeAll` section that loads a model. 16-19ms for tests execution, and ~3seconds for model loading. I will implement some sort of caching to speed up it a bit. Is 1.5 second of model load will be considered as acceptable? WDYT?

**Comment by volodya-lombrozo** [comment]
@h1alexbel Yes, it would be great. `1.5` seconds is acceptable. Please, go ahead.

**Comment by h1alexbel** [comment]
@volodya-lombrozo updated pull request, take a look again, please

File: src/test/java/com/github/lombrozo/testnames/rules/ml/ModelSourceFileSystemTest.java [hunk 1]
```diff
@@ -25,27 +25,41 @@
...
 @Disabled
 final class ModelSourceFileSystemTest {
 
+    /**
+     * Cache.
+     */
+    private static final AtomicReference<POSModel> CACHE = new AtomicReference<>();
+
+    @BeforeAll
+    static void setUp() throws IOException {
```
Comment (volodya-lombrozo) [inline src/test/java/com/github/lombrozo/testnames/rules/ml/ModelSourceFileSystemTest.java:50]: @h1alexbel As I have understood your idea, you are trying to warm-up the cache before the usage in all the tests. I believe you don't need to do this. The first call to the `mode.model()` in the first test should do all the work. The rest tests will reuse cached model.

(At least, it must be so. If I'm wrong, please, correct me.)

File: src/test/java/com/github/lombrozo/testnames/rules/ml/ModelSourceFileSystemTest.java [hunk 2]
```diff
@@ -25,27 +25,41 @@
...
+        if (ModelSourceFileSystemTest.CACHE.get() == null) {
+            model.model();
+        }
+    }
+
     @Test
     void loadsFromFileSystem(@TempDir final Path temp) throws IOException {
         final Path path = temp.resolve("model.bin");
-        new ModelSourceInternet().model().serialize(path);
+        ModelSourceFileSystemTest.CACHE.get().serialize(path);
```
Comment (volodya-lombrozo) [inline src/test/java/com/github/lombrozo/testnames/rules/ml/ModelSourceFileSystemTest.java:62]: @h1alexbel The usage of `public static` variables is considered as an anti-pattern. You can read more about it here:

https://stackoverflow.com/questions/51894749/best-practice-to-access-static-object-public-static-or-private-static-with

https://www.yegor256.com/2015/07/06/public-static-literals.html

File: src/test/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        this.origin = orgn;
+        this.memory = cache;
+    }
+
+    @Override
+    public POSModel model() throws IOException {
+        final POSModel model;
+        final File cached = Paths.get("src/test/resources/ml/cached.bin").toFile();
+        final POSModel memorized = this.memory.get();
+        if (memorized != null) {
```
Comment (volodya-lombrozo) [inline src/test/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:71]: @h1alexbel I would highly recommend to use `Sticky` object from `cactoos` instead your own caching implementation.
Here is the detailed guide how to implement caching with it:
https://www.yegor256.com/2017/10/17/lazy-loading-caching-sticky-cactoos.html

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]
@h1alexbel Thank you for the contribution! I left several significant comments. Would you be able to address them, please?

File: src/test/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        final AtomicReference<POSModel> cache
+    ) {
+        this.origin = orgn;
+        this.memory = cache;
+    }
+
+    @Override
+    public POSModel model() throws IOException {
+        final POSModel model;
+        final File cached = Paths.get("src/test/resources/ml/cached.bin").toFile();
```
[the remark this answers is not in the document]
Comment (h1alexbel) [reply src/test/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:69]: I like it

**Comment by h1alexbel** [comment]
@volodya-lombrozo updated pull request. Take a look again, please

File: src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        final ModelSource orgn,
+        final File ccd
+    ) {
+        this.origin = orgn;
+        this.cached = ccd;
+    }
+
+    @Override
+    public POSModel model() throws Exception {
+        return new Sticky<>(
```
Comment (volodya-lombrozo) [inline src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:86]: @h1alexbel `Sticky` objects keeps the return value in the RAM.  So, it's better to keep it as a field and reuse each time when we call `model()`. Without this, `Sticky` object usage doesn't make sense.  All of the rest is good!

In other words we will have such levels of caching:

- RAM (`Sticky` object)
- FS (You perfectly implemented it)
- INTERNET (`ModelSourseInternet`)

File: pom.xml [hunk 6]
```diff
@@ -435,5 +436,30 @@ SOFTWARE.
         </plugins>
       </build>
     </profile>
+    <profile>
+      <!-- Deep Testing: https://www.yegor256.com/2023/08/22/fast-vs-deep-testing.html -->
+      <id>deep</id>
```
Comment (volodya-lombrozo) [inline pom.xml:441]: @h1alexbel It's a good thing actually! However I suggest to keep in this PR only the changes relevant to the #101 issue. We can add this `deep` profile in a separate PR. What do you think?

**Review by volodya-lombrozo — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 6 above)
Comment (h1alexbel) [reply pom.xml:441]: Agree

File: src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java (hunk 5 above)
Comment (h1alexbel) [reply src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:86]: @volodya-lombrozo do you mean to pass it in constructor, compute it there, and reuse in `model()`?

Comment (h1alexbel) [reply src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:86]: @volodya-lombrozo as far I understand:

```java
public CachedModelSource(
        final ModelSource orgn
    ) {
        this(
            orgn,
            "src/test/resources/ml/cached.bin",
            new Scalar<POSModel>() {
            @Override
            public POSModel value() throws Exception {
                return new Sticky<>(
                    () -> {
                        final POSModel model;
                        if (this.cached.exists()) {
                            model = new POSModel(this.cached);
                        } else {
                            if (!this.cached.toPath().isAbsolute() && !this.cached.exists()) {
                                Files.createDirectory(
                                    Paths.get(
                                        this.cached.getParent()
                                    )
                                );
                            }
                            model = this.origin.model();
                            model.serialize(this.cached);
                        }
                        return model;
                    }
                ).value();
            }
        });
    }
```

Comment (volodya-lombrozo) [reply src/main/java/com/github/lombrozo/testnames/rules/ml/CachedModelSource.java:86]: @h1alexbel Yes, something like this! 

Please, take a look how we did it in EO: https://github.com/objectionary/eo/blob/master/eo-maven-plugin/src/main/java/org/eolang/maven/hash/ChCached.java

**[Optional]** This technique might be rather useful in your case: https://www.yegor256.com/2021/08/04/prestructors.html

**Review by volodya-lombrozo — APPROVED** [review]
@h1alexbel Everything is perfectly well. Thank you for the contribution.

**Comment by volodya-lombrozo** [comment]
@rultor merge

**Pull request merged**